feat: table plugin

zoli opened the pull request:
Coming from https://github.com/AppFlowy-IO/AppFlowy/pull/2120.

- New `Node` types 'table' and 'table/cell'. The 'table/cell' node for now contains single `TextNode`. All the data is kept in node attributes. Data related to 'table' like number of rows and columns are in 'table' node attributes and data related to each table cell is in 'table/cell' node attribute (cell height, position of cell its row and column).  The `TextNode` in table cell supports selection menu and toolbar but not all of them work inside the table cell (some of them are not applicable in table cell).
- Added `isApplicable` to `ContextMenuItem` so we can have context-base context menu
- Modified `backspaceEventHandler` so it won't move from one table cell to another cell and also handling backspace when the selection includes part of table cells and none table related nodes.
- Modified `_getNodeInOffset` so selection also works when we have multiple nodes in a single line (same y position). I will add more info on this in another comment.
- Made `ColorPicker` more customizable and reusable for other than toolbar service.

zoli committed:
feat: table plugin requisites

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/AppFlowy-IO/appflowy-editor?pullRequest=62) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/AppFlowy-IO/appflowy-editor?pullRequest=62) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/AppFlowy-IO/appflowy-editor?pullRequest=62) it.</sub>

zoli committed:
feat: moved table plugin to appflowy-editor

zoli committed:
fix: flutter analyze

zoli committed:
fix: dont use nested maps in node attributes

rust backend (flowy-document) doesn't support it.

zoli committed:
feat: add table to appflowy editor default selection menu

zoli committed:
feat table markdown encoder

Xazin commented on lib/src/render/color_menu/color_picker.dart:
```diff
@@ -105,21 +102,27 @@ class _ColorPickerState extends State<ColorPicker> {
-          .map((e) => _buildColorItem(type, e, e.colorHex == selectedColor))
+      children: colorOptionList.colorOptions
+          .map(
+            (e) => _buildColorItem(
+              colorOptionList.onSubmittedAction,
+              e,
+              e.colorHex == colorOptionList.selectedColorHex,
```
```suggestion
            (colorOption) => _buildColorItem(
              colorOptionList.onSubmittedAction,
              colorOption,
              colorOption.colorHex == colorOptionList.selectedColorHex,
```
Makes it easier to read, as you don't have to "guess" what `e` refers to, the default `e` refers to `element`, which can be anything.

Xazin commented on lib/src/render/table/table_action.dart:
```diff
@@ -0,0 +1,218 @@
+    );
+  }
+
+  for (var i = row + 1; i < rowsLen; i++) {
+    for (var j = 0; j < colsLen; j++) {
+      transaction.updateNode(getCellNode(tableNode, j, i)!, {
+        'colPosition': j,
+        'rowPosition': i + 1,
+      });
+    }
+  }
```
I see varying versions of the same nested for loop,

Maybe possible to consolidate it into a private method?

Xazin commented on lib/src/render/table/table_cell_node_widget.dart:
```diff
@@ -0,0 +1,75 @@
+        if (n.attributes['backgroundColor'] == null) {
+          return null;
+        }
+        var colorInt = int.tryParse(n.attributes['backgroundColor']);
+        return colorInt != null ? Color(colorInt) : null;
```
```suggestion
        final colorInt = int.tryParse(n.attributes['backgroundColor']);
        return colorInt != null ? Color(colorInt) : null;
```
Can be final

Xazin commented on lib/src/render/table/table_cell_node_widget.dart:
```diff
@@ -0,0 +1,75 @@
+        var colorInt = int.tryParse(n.attributes['backgroundColor']);
+        return colorInt != null ? Color(colorInt) : null;
+      }),
+      width: double.infinity,
```
Is this actually needed?

Xazin commented on lib/src/render/table/table_col_border.dart:
```diff
@@ -0,0 +1,76 @@
+        onHorizontalDragStart: (_) => setState(() => _borderDragging = true),
+        onHorizontalDragEnd: (_) => setState(() => _borderDragging = false),
+        onHorizontalDragUpdate: (DragUpdateDetails details) {
+          RenderBox box =
+              _borderKey.currentContext?.findRenderObject() as RenderBox;
```
```suggestion
          final RenderBox box =
              _borderKey.currentContext?.findRenderObject() as RenderBox;
```
Is not reassigned

Xazin commented on lib/src/render/table/table_col_border.dart:
```diff
@@ -0,0 +1,76 @@
+        onHorizontalDragUpdate: (DragUpdateDetails details) {
+          RenderBox box =
+              _borderKey.currentContext?.findRenderObject() as RenderBox;
+          Offset pos = box.localToGlobal(Offset.zero);
+          double colsHeight = widget.tableNode.colsHeight;
```
```suggestion
          final Offset pos = box.localToGlobal(Offset.zero);
          final double colsHeight = widget.tableNode.colsHeight;
```
These are not reassigned either as far as I can tell

Xazin commented on lib/src/render/table/table_col_border.dart:
```diff
@@ -0,0 +1,76 @@
+            return;
+          }
+
+          final w = widget.tableNode.getColWidth(widget.colIdx);
```
```suggestion
          final colWidth = widget.tableNode.getColWidth(widget.colIdx);
```
`w` is very ambiguous

Xazin commented on lib/src/render/table/table_context_menu_item.dart:
```diff
@@ -0,0 +1,217 @@
+                colorOptions: generateBackgroundColorOptions(editorState),
+                onSubmittedAction: (color) {
+                  final transaction = editorState.transaction;
+                  var c = color == node.attributes['backgroundColor']
```
```suggestion
                  final c = color == node.attributes['backgroundColor']
```
Not sure what `c` should stand for here? maybe use eg. `is` or `has` like `final hasBgColor` for readability

Xazin commented on lib/src/render/table/table_context_menu_item.dart:
```diff
@@ -0,0 +1,217 @@
+  // workaround: SelectionService has been released after hot reload.
+  final isSelectionDisposed =
+      _editorState?.service.selectionServiceKey.currentState == null;
+  if (isSelectionDisposed) {
+    return;
+  }
+  if (_editorState?.service.selectionService.currentSelection.value == null) {
+    return;
+  }
```
```suggestion
  if (isSelectionDisposed || _editorState?.service.selectionService.currentSelection.value == null) {
    return;
  }
```
Consolidate

zoli committed:
refactor: apply @Xazin review suggestions

zoli committed:
refactor: merge main and fix color picker conflicts

zoli committed:
refactor: merge main, last merge i made mistake

zoli committed:
feat: table markdown decoder

zoli committed:
fix: use specific backgroun generator for table

zoli commented:
Implemented markdown decode/encode feature for table.

LucasXu0 committed:
feat: table plugin

zoli committed:
style: adapt new structure

zoli committed:
fix: table columns border height

zoli committed:
fix: adapt new style key commands

zoli committed:
style(merge): merge migrate_table_plugin branch

zoli committed:
fix: remove table context menu

zoli commented:
Just wanted to note the main description of this PR is outdated https://github.com/AppFlowy-IO/appflowy-editor/pull/62#issue-1663045067.

zoli committed:
test: adapt to main branch changes

zoli committed:
style: merge branch 'main' into feat/table-plugin

* main:
  fix: error occurs when applying divider on Mobile (#252)
  feat: text and background color in mobile toolbar (#233)
  fix: extend attribute keys shouldn't be sliced (#248)
  feat: support customizing text attribute key and rendering (#244)

zoli committed:
fix: backspace handling regarding table

zoli committed:
fix: flutter analyze

zoli committed:
fix: enter when table cell is bullet list

zoli committed:
fix: selection bug on large table

zoli committed:
feat: row/col action menu

zoli committed:
style: merge branch 'main' into feat/table-plugin

* main:
  feat: support uploading images from local files  (#232)
  fix: desktop tooltips incorrect (#259)
  fix: add underline syntax parser (#256)
  chore: add reload function into editor state
  fix: ignore the same value update in input service (#257)
  Revert "feat: support mobile IME (#255)"
  feat: support item count filter in selection menu results  (#254)
  feat: support mobile IME (#255)
  feat: replace DeltaTextInputClient with TextInputClient (#253)

zoli committed:
fix: row action handler positioning

zoli committed:
refactor: merge row and col handler widget

zoli committed:
refactor: add size and color to handler icon

LucasXu0 commented:
@zoli Is it ready to review and merge?

zoli commented:
> @zoli Is it ready to review and merge?

Yes. The design and maybe some UX corners could improve, But I think enhancing them incrementally is better.

zoli committed:
style(merge): merge main

zoli committed:
style(merge): merge main

zoli committed:
style(merge): merge main

zoli committed:
style(merge): merge main and resolve conflicts

annieappflowy commented:
I can't wait to try this. When will this feature be released?

zoli commented:
> I can't wait to try this. When will this feature be released?

In my last discussion with @LucasXu0, he said we should move the plugin to Appflowy repo.

LucasXu0 committed:
Merge branch 'main' into feat/table-plugin

LucasXu0 committed:
Merge branch 'main' into feat/table-plugin

zoli committed:
Merge branch 'main' into feat/table-plugin

* main:
  fix: refactor _onSelectionChange method to update _showPlaceholder state more efficiently
  fix: typo (#408)
  Fix selected nodes (#407)
  fix: miss the nested node when copying (#406)

zoli committed:
feat: added padding and action wrapper

zoli committed:
feat: enable customizing icons and etc

zoli committed:
feat: table menu builder

LucasXu0 committed:
Merge remote-tracking branch 'origin/main' into feat/table-plugin

zoli committed:
style: merge branch 'main' into feat/table-plugin

* main:
  chore: comment suggesting use of deprecated method (#417)
  fix: replace matches on the same node (#418)

zoli committed:
fix: table getBlockRect

zoli committed:
refactor: table action button design
